packages/compiler/core/checker.ts

```diff
@@ -1789,6 +1960,1115 @@ export function createChecker(program: Program): Checker {
+    evalContext = createEvalContext(node);
+    for (const [i, param] of node.parameters.entries()) {
+      if (!args[i]) {
+        throw new Error("need argument for parameter " + node.parameters[i]);
+      }
+
+      let argVal = args[i];
+      let typeVal;
+
+      if (typeof argVal === "number" || typeof argVal === "string" || typeof argVal === "boolean") {
+        typeVal = valueToLiteralType(argVal);
+      } else {
+        typeVal = argVal;
+      }
+
+      evalContext.locals.set(param.id.sv, typeVal);
+    }
+    const clone = cloneType(target);
+
+    clone.projectionSource = target;
+    evalContext.locals.set("self", clone);
+    let lastVal: TypeOrReturnRecord = voidType;
+    for (const item of node.body) {
+      lastVal = evalProjectionNode(item);
+      if (lastVal.kind === "Return") {
+        break;
+      }
+    }
+
+    if (topLevelProjection) {
+      currentProjectionDirection = undefined;
+    }
+    evalContext = originalContext;
+
+    if (lastVal.kind === "Return") {
+      return lastVal.value;
+    } else {
+      return clone;
+    }
+  }
+
+  function evalProjectionExpressionStatement(node: ProjectionExpressionStatement) {
+    return evalProjectionNode(node.expr);
+  }
+
+  function evalProjectionCallExpression(node: ProjectionCallExpressionNode) {
+    const target = evalProjectionNode(node.target);
+
+    if (!target) throw new Error("target undefined");
+    const args = [];
+    for (const arg of node.arguments) {
+      args.push(evalProjectionNode(arg));
+    }
+
+    if (target.kind !== "Function") {
+      throw new Error("Can't call non-function, got type " + target.kind);
+    }
+
+    return target.call(...args);
+  }
+
+  function evalProjectionMemberExpression(node: ProjectionMemberExpressionNode): Type {
+    const base = evalProjectionNode(node.base);
+    const member = node.id.sv;
+
+    switch (base.kind) {
+      case "Namespace":
+        const sym = base.node.exports!.get(member) as TypeSymbol;
+        if (sym) {
+          const links = getSymbolLinks(sym);
+          return links.declaredType || links.type || errorType;
+        } else {
+          return errorType;
+        }
+      case "Model":
+        switch (member) {
+          case "projectionSource":
+            return base.projectionSource?.projectionSource || voidType;
+          case "properties":
+            const props: ObjectType = createType({
+              kind: "Object",
+              properties: {},
+            });
+
+            props.properties.forEach = functionTypeFromFunction(
+              (_: Program, block: FunctionType) => {
+                const props = Array.from(base.properties.values());
+                props.forEach((p) => block.call(p));
+                return voidType;
+              }
+            );
+
+            return props;
+          case "renameProperty":
+            return functionTypeFromFunction((_: Program, oldNameT: Type, newNameT: Type) => {
```

Comment:
It's tempting to refactor this function in particular, but renaming is subtly different across the different types. For now, I've opted to duplicate some logic.